### PR TITLE
Generate CIFTI and TSV versions of coverage, timeseries, and correlation files

### LINKED ISSUE #555: Supplement parcellated time series and connectivity CIFTI files with TSVs
## Summary

The atlas time series and connectivity matrices are currently stored as CIFTI files. We should change them to TSV files.

## Additional details

This stems from @kahinimehta's comment in https://github.com/PennLINC/xcp_d/issues/517#issuecomment-1265808370.
I've split it off into a separate issue because this is a file format change, rather than just a naming change.

### LINKED ISSUE #777: Ensure that zeroed-out time series/correlation coefficients are set to NaNs
## Summary

`wb_command -cifti-correlate` produces ones in the diagonal, even for time series of all zeros. There may be other edge cases where the zeroed-out time series don't end up as NaNs.

### PR DESCRIPTION
**Pull request by tsalo**
Closes #555 and closes #777. Supercedes #559 and #783.

## Changes proposed in this pull request
- When using the `--cifti` flag, generate and write out tsv versions of the coverage, timeseries, and correlation files for each atlas. 
    - The tsv versions will have the node labels from the atlases' associated node labels files, rather than the node labels that are present within the CIFTI atlases. However, the CIFTI versions of these files will still have the original node names.
- Remove unused `CiftiPrepareForParcellation` and `CiftiCorrelation` interfaces.

### CODE REVIEW

File: xcp_d/interfaces/connectivity.py [hunk 1]
```diff
@@ -119,23 +140,321 @@ def _run_interface(self, runtime):
...
+                new_parcel_coverage[label_row] = parcel_coverage[row]
+
+            parcel_coverage = new_parcel_coverage
+
+        # The time series file is tab-delimited, with node names included in the first row.
+        timeseries_df = pd.DataFrame(data=timeseries_arr, columns=node_labels)
+        correlations_df = timeseries_df.corr()
+        coverage_df = pd.DataFrame(data=parcel_coverage, index=node_labels)
+
+        timeseries_df.to_csv(self._results["timeseries"], sep="\t", na_rep="NaN", index=False)
```
Comment (tsalo) [inline xcp_d/interfaces/connectivity.py:186]: I started writing NaNs as "NaN" (instead of empty cells) in the tsv file. There might be something better (e.g., "n/a", as BIDS prescribes for missing data). Does anyone have a strong opinion on this?

Comment (tsalo) [reply xcp_d/interfaces/connectivity.py]: Actually switching to "n/a".

**Comment by tsalo** [comment]
@mattcieslak I think this is finally ready to review.

File: xcp_d/interfaces/connectivity.py [hunk 2]
```diff
@@ -79,27 +90,37 @@ def _run_interface(self, runtime):
...
+            resampling_target=None,  # they should be in the same space/resolution already
         )
         n_voxels_in_masked_parcels = sum_masker_masked.fit_transform(atlas_img_bin)
         n_voxels_in_parcels = sum_masker_unmasked.fit_transform(atlas_img_bin)
         parcel_coverage = np.squeeze(n_voxels_in_masked_parcels / n_voxels_in_parcels)
         coverage_thresholded = parcel_coverage < coverage_threshold
 
-        n_nodes = coverage_thresholded.size
+        n_nodes = len(node_labels)
+        n_found_nodes = coverage_thresholded.size
```
Comment (mattcieslak) [inline xcp_d/interfaces/connectivity.py:114]: I've never seen .size used this way before.

File: xcp_d/interfaces/connectivity.py [hunk 3]
```diff
@@ -119,23 +140,321 @@ def _run_interface(self, runtime):
...
-        time_series = masker.fit_transform(filtered_file)
+        timeseries_arr = masker.fit_transform(filtered_file)
+        assert timeseries_arr.shape[1] == n_found_nodes
 
         # Apply the coverage mask
-        time_series[:, coverage_thresholded] = 0
+        timeseries_arr[:, coverage_thresholded] = np.nan
+
+        # Region indices in the atlas may not be sequential, so we map them to sequential ints.
+        seq_mapper = {idx: i for i, idx in enumerate(node_labels_df.index.tolist())}
```
Comment (mattcieslak) [inline xcp_d/interfaces/connectivity.py:158]: is there a guarantee that they're already in order? Would it make sense to have `enumerate(sorted(node_labels_df.index.tolist()))`

File: xcp_d/interfaces/connectivity.py [hunk 4]
```diff
@@ -119,23 +140,321 @@ def _run_interface(self, runtime):
...
+    def _run_interface(self, runtime):
+        coverage_threshold = self.inputs.min_coverage
+        data_file = self.inputs.data_file
+        atlas_file = self.inputs.atlas_file
+        pscalar_file = self.inputs.parcellated_atlas
+        atlas_labels = self.inputs.atlas_labels
+
+        cifti_intents = get_cifti_intents()
+
+        assert data_file.endswith(".dtseries.nii"), data_file
```
Comment (mattcieslak) [inline xcp_d/interfaces/connectivity.py:262]: won't `mandatory=True` enforce that data_file is defined?

File: xcp_d/interfaces/connectivity.py [hunk 5]
```diff
@@ -119,23 +140,321 @@ def _run_interface(self, runtime):
...
+        label_axis = atlas_img.header.get_axis(0)
+        parcels_axis = pscalar_img.header.get_axis(1)
+        atlas_label_mapper = label_axis.label[0]
+        atlas_label_mapper = {k: v[0] for k, v in atlas_label_mapper.items()}
+
+        if 0 in atlas_label_mapper.keys():
+            atlas_label_mapper.pop(0)
+
+        # Explicitly remove label corresponding to background (index=0), if present.
+        if 0 in node_labels_df.index:
```
Comment (mattcieslak) [inline xcp_d/interfaces/connectivity.py:292]: may want to add a warning here too

**Review by mattcieslak — COMMENTED** [review]
a massive undertaking, looks good to me. Just minor comments/questions

File: xcp_d/interfaces/connectivity.py (hunk 4 above)
Comment (tsalo) [reply xcp_d/interfaces/connectivity.py:262]: True, but I wanted to make sure that only dtseries.nii files were fed into this interface.

File: xcp_d/tests/test_workflows_connectivity.py [hunk 6]
```diff
@@ -56,59 +56,76 @@ def test_nifti_conn(fmriprep_with_freesurfer_data, tmp_path_factory):
...
+
+    # Parcels with <50% coverage should have NaNs
+    assert np.array_equal(np.squeeze(coverage_arr) < 0.5, np.isnan(np.diag(correlations_arr)))
 
     # Now let's get the ground truth. First, we should locate the atlas
     atlas_file = os.path.join(
         connectivity_wf.base_dir,
-        "connectivity_wf/warp_atlases_to_bold_space/mapflow/_warp_atlases_to_bold_space3",
-        "Schaefer2018_400Parcels_17Networks_order_FSLMNI152_2mm_trans.nii.gz",
+        "connectivity_wf/warp_atlases_to_bold_space/mapflow/_warp_atlases_to_bold_space9",
```
[the remark this answers is not in the document]
Comment (tsalo) [reply xcp_d/tests/test_workflows_connectivity.py:87]: Me too!

File: xcp_d/workflows/outputs.py [hunk 7]
```diff
@@ -198,6 +202,67 @@ def init_writederivatives_wf(
...
+            suffix="conmat",
+            extension=".tsv",
+        ),
+        name="ds_correlations",
+        run_without_submitting=True,
+        mem_gb=1,
+        iterfield=["atlas", "in_file"],
+    )
+
+    # fmt:off
```
[the remark this answers is not in the document]
Comment (tsalo) [reply xcp_d/workflows/outputs.py:249]: Yeah, this is how I tell black not to reformat the workflow connections, which is where it tends to make nipype code terrible.

File: xcp_d/workflows/connectivity.py [hunk 8]
```diff
@@ -182,6 +223,15 @@ def init_nifti_functional_connectivity_wf(
         mem_gb=mem_gb,
     )
 
+    # fmt:off
+    workflow.connect([
+        (inputnode, matrix_plot, [("clean_bold", "in_file")]),
+        (atlas_name_grabber, matrix_plot, [["atlas_names", "atlas_names"]]),
```
[the remark this answers is not in the document]
Comment (tsalo) [reply xcp_d/workflows/connectivity.py:229]: Oh good catch!

[the remark this answers is not in the document]
Comment (tsalo) [reply xcp_d/workflows/connectivity.py]: Fixed.

File: xcp_d/interfaces/connectivity.py [hunk 9]
```diff
@@ -26,12 +30,13 @@ class _NiftiConnectInputSpec(BaseInterfaceInputSpec):
...
     mask = File(exists=True, mandator=True, desc="brain mask file")
     atlas = File(exists=True, mandatory=True, desc="atlas file")
+    atlas_labels = File(exists=True, mandatory=True, desc="atlas labels file")
 
 
 class _NiftiConnectOutputSpec(TraitedSpec):
-    time_series_tsv = File(exists=True, mandatory=True, desc=" time series file")
-    fcon_matrix_tsv = File(exists=True, mandatory=True, desc=" time series file")
-    parcel_coverage_file = File(exists=True, mandatory=True, desc="Parcel-wise coverage file.")
+    timeseries = File(exists=True, mandatory=True, desc=" time series file")
```
[the remark this answers is not in the document]
Comment (tsalo) [reply xcp_d/interfaces/connectivity.py]: Done!

File: xcp_d/interfaces/connectivity.py [hunk 10]
```diff
@@ -49,24 +54,30 @@ def _run_interface(self, runtime):
         filtered_file = self.inputs.filtered_file
         mask = self.inputs.mask
         atlas = self.inputs.atlas
+        atlas_labels = self.inputs.atlas_labels
+
+        node_labels_df = pd.read_table(atlas_labels, index_col="index")
+
+        # Explicitly remove label corresponding to background (index=0), if present.
+        if 0 in node_labels_df.index:
```
[the remark this answers is not in the document]
Comment (tsalo) [reply xcp_d/interfaces/connectivity.py:63]: Good point. I've added a warning.

File: xcp_d/interfaces/connectivity.py (hunk 3 above)
Comment (tsalo) [reply xcp_d/interfaces/connectivity.py:163]: That's a good point. I just added something to work around this. I implemented it a little differently, but it should accomplish the same thing.

File: xcp_d/interfaces/connectivity.py (hunk 5 above)
Comment (tsalo) [reply xcp_d/interfaces/connectivity.py:298]: Done.

File: xcp_d/tests/test_workflows_connectivity.py [hunk 11]
```diff
@@ -56,59 +56,76 @@ def test_nifti_conn(fmriprep_with_freesurfer_data, tmp_path_factory):
...
-    # Parcels with <50% coverage should have NaNs
-    # We know that 10 of the parcels in the 400-parcel Schaefer are flagged
-    assert np.sum(np.isnan(np.diag(xcp_array))) == 10
+    # Drop missing parcels
+    correlations_arr = correlations_arr[idx_in_atlas, :]
+    correlations_arr = correlations_arr[:, idx_in_atlas]
+    assert correlations_arr.shape == (n_nodes_in_atlas, n_nodes_in_atlas)
+
+    # The masker.labels_ attribute only contains the labels that were found
+    assert idx_not_in_atlas.size == 2
```
[the remark this answers is not in the document]
Comment (mattcieslak) [reply xcp_d/tests/test_workflows_connectivity.py:114]: and it's just in the 1000 parcel cifti? I'll open an issue

[the remark this answers is not in the document]
Comment (mattcieslak) [reply xcp_d/tests/test_workflows_connectivity.py:114]: nope, not going to be fixed: 

https://github.com/netneurolab/neuromaps/issues/66
https://github.com/ThomasYeoLab/CBIG/issues/43
https://github.com/ThomasYeoLab/CBIG/issues/10

**Comment by tsalo** [comment]
Going to merge. The file changes will be a drop in the bucket of the git history.
